samples/nrf9160/nrf_cloud_rest_device_message/src/main.c

```diff
@@ -0,0 +1,379 @@
+
+	/* Connect to LTE */
+	err = connect_to_lte();
+
+	if (err) {
+		LOG_ERR("Failed to connect to cellular network: %d", err);
+		return err;
+	}
+
+	/* Wait until we know what time it is (necessary for JSON Web Token generation) */
+	err = modem_time_wait();
+
+	if (err) {
+		LOG_ERR("Modem failed to acquire real time: %d", err);
+		return err;
+	}
+
+	/* Perform JITP if enabled and requested */
+	if (IS_ENABLED(CONFIG_REST_DEVICE_MESSAGE_DO_JITP) && jitp_requested) {
+		err = do_jitp();
+
+		if (err) {
+			LOG_ERR("Failed to perform JITP, %d", err);
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static void offer_jitp(void)
+{
+	int ret;
+
+	jitp_requested = false;
+
+	k_sem_reset(&button_press_sem);
+
+	LOG_INF("---> Press button %d to request just-in-time provisioning", BTN_NUM);
+	LOG_INF("     Waiting %d seconds...", JITP_REQ_WAIT_SEC);
+
+	ret = k_sem_take(&button_press_sem, K_SECONDS(JITP_REQ_WAIT_SEC));
+	if (ret == 0) {
+		jitp_requested = true;
+		LOG_INF("JITP will be performed after network connection is obtained");
+	} else {
+		if (ret != -EAGAIN) {
+			LOG_ERR("k_sem_take error: %d", ret);
+		}
+
+		LOG_INF("JITP will be skipped");
+	}
+}
+
+int init(void)
+{
+	/* Init the LED */
+	int err = init_led();
+
+	if (err) {
+		LOG_ERR("LED initialization failed");
+		return err;
+	}
+
+	/* Init modem */
+	int modem_lib_init_result = nrf_modem_lib_init(NORMAL_MODE);
+
+	if (modem_lib_init_result) {
+		LOG_ERR("Failed to initialize modem library: 0x%X", modem_lib_init_result);
+		return -EFAULT;
+	}
+
+	/* Init the button */
+	err = dk_buttons_init(button_handler);
+	if (err) {
+		LOG_ERR("Failed to initialize button: error: %d", err);
+		return err;
+	}
+
+	return 0;
+}
+
+static void error_reboot(void)
+{
+	LOG_INF("Rebooting in 30s..");
+	(void)nrf_cloud_rest_disconnect(&rest_ctx);
+	(void)lte_lc_deinit();
+	k_sleep(K_SECONDS(30));
+	sys_reboot(SYS_REBOOT_COLD);
+}
+
+int setup(void)
+{
+	/* Initialize libraries and hardware */
+	int err = init();
+
+	if (err) {
+		LOG_ERR("Initialization failed.");
+		return err;
+	}
+
+	/* Ask user if JITP via REST is desired */
+	if (IS_ENABLED(CONFIG_REST_DEVICE_MESSAGE_DO_JITP)) {
+		offer_jitp();
+	}
+
+	/* Initiate Connection */
+	err = setup_connection();
+	if (err) {
+		LOG_ERR("Connection set-up failed.");
+		return err;
+	}
+
+	return 0;
+}
+
+void main(void)
+{
+	LOG_INF("nRF Cloud REST Device Message Sample, version: %s",
+		CONFIG_REST_DEVICE_MESSAGE_SAMPLE_VERSION);
+
+	int err = setup();
+
+	if (err) {
+		LOG_ERR("Setup failed, stopping.");
+		return;
+	}
+
+	/* Send off a hello world message! */
+	/* This does not match a predefined schema. But that's not a problem! */
+	err = send_message("{\"sample_message\":"
+					    "\"Hello World, from the REST Device Message Sample!\"}");
+
+	if (err) {
+		LOG_ERR("Hello World failed, stopping.");
+		return;
+	}
+
+	while (1) {
+		err = send_message_on_button();
+
+		if (err) {
+			error_reboot();
```

Comment:
Is the reboot the only way to recover?  Seems heavy-handed.

Reply:
We could try resend on failure, but I don't think the added complexity would be appropriate for a sample like this,

The other alternative would be to simply not recover (just return out of main). Maybe that would be better?

Reply:
I'm sensitive to the overuse of reboot as a recovery mechanism in our samples in general.   Customers will see this and assume that is all one can do, and then complain later that their device reboots all the time.

Reply:
I'd let it retry the button message the next time the user pressed the button, and only log the transmission failure.

Reply:
That's a good idea -- message retry essentially for free in terms of sample complexity. I'll do that!